runtime/src/accounts_db.rs

```diff
@@ -1609,20 +1612,44 @@ impl AccountsDb {
 
     // Construct a vec of pubkeys for cleaning from:
     //   uncleaned_pubkeys - the delta set of updated pubkeys in rooted slots from the last clean
-    //   zero_lamport_pubkeys - set of all alive pubkeys containing 0-lamport updates
+    //   dirty_stores - set of stores which had accounts removed
     fn construct_candidate_clean_keys(
         &self,
         max_clean_root: Option<Slot>,
         timings: &mut CleanKeyTimings,
     ) -> Vec<Pubkey> {
         let mut zero_lamport_key_clone = Measure::start("zero_lamport_key");
-        let pubkeys = self.accounts_index.zero_lamport_pubkeys().clone();
+        let max_slot = max_clean_root.unwrap_or_else(|| self.accounts_index.max_root());
+        let dirty_stores: Vec<(Slot, Arc<AccountStorageEntry>)> = self
+            .dirty_stores
+            .iter()
+            .filter_map(|entry| {
+                if entry.key().0 <= max_slot {
+                    Some((entry.key().0, entry.value().clone()))
+                } else {
+                    None
+                }
+            })
+            .collect();
+        self.dirty_stores
+            .retain(|(slot, _store_id), _store_ref| *slot > max_slot);
```

Comment:
Can we avoid the `filter_map` iteration above if we add to the vector in the `retain` here?

Reply:
yes. updated!